Userland/Libraries/LibCompress/Brotli.cpp

```diff
@@ -0,0 +1,905 @@
+        }
+        mtf[0] = value;
+    }
+}
+
+ErrorOr<void> BrotliDecompressionStream::read_context_map(size_t number_of_codes, Vector<u8>& context_map, size_t context_map_size)
+{
+    bool use_run_length_encoding = TRY(m_input_stream.read_bit());
+    size_t run_length_encoding_max = 0;
+    if (use_run_length_encoding) {
+        run_length_encoding_max = 1 + TRY(m_input_stream.read_bits(4));
+    }
+    // dbgln("run_length_encoding_max: {}", run_length_encoding_max);
+
+    BrotliDecompressionStream::CanonicalCode code;
+    TRY(read_prefix_code(code, number_of_codes + run_length_encoding_max));
+
+    size_t i = 0;
+    while (i < context_map_size) {
+        // dbgln("i: {}", i);
+        size_t symbol = code.read_symbol(m_input_stream).release_value();
+        // dbgln("symbol: {}", symbol);
+
+        if (symbol <= run_length_encoding_max) {
+            size_t repeat_base = 1 << symbol;
+            size_t repeat_additional = TRY(m_input_stream.read_bits(symbol));
+            size_t repeat_count = repeat_base + repeat_additional;
+            // dbgln("repeat_count: {}", repeat_count);
+            while (repeat_count--) {
+                context_map.append(0);
+                i++;
+            }
+        } else {
+            size_t value = symbol - run_length_encoding_max;
+            context_map.append(value);
+            i++;
+        }
+    }
+
+    // for (i = 0; i < context_map.size(); i++) {
+    //     dbgln("{}: {}", i, context_map[i]);
+    // }
+
+    bool inverse_move_to_front = TRY(m_input_stream.read_bit());
+    if (inverse_move_to_front) {
+        // dbgln("inverse move to front");
+        inverse_move_to_front_transform(context_map.span());
+    }
+
+    // for (i = 0; i < context_map.size(); i++) {
+    //     dbgln("{}: {}", i, context_map[i]);
+    // }
+
+    return {};
+}
+
+ErrorOr<void> BrotliDecompressionStream::read_block_configuration(Block& block)
+{
+    size_t blocks_of_type = TRY(read_variable_length());
+
+    block.type = 0;
+    block.type_previous = 1;
+    block.number_of_types = blocks_of_type;
+
+    if (blocks_of_type == 1) {
+        block.length = 16 * MiB;
+    } else {
+        TRY(read_prefix_code(block.type_code, 2 + blocks_of_type));
+        TRY(read_prefix_code(block.length_code, 26));
+        TRY(block_update_length(block));
+    }
+
+    return {};
+}
+
+ErrorOr<void> BrotliDecompressionStream::block_update_length(Block& block)
+{
+    size_t const block_length_code_base[26] { 1, 5, 9, 13, 17, 25, 33, 41, 49, 65, 81, 97, 113, 145, 177, 209, 241, 305, 369, 497, 753, 1265, 2289, 4337, 8433, 16625 };
+    size_t const block_length_code_extra[26] { 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6, 7, 8, 9, 10, 11, 12, 13, 24 };
+
+    size_t symbol = TRY(block.length_code.read_symbol(m_input_stream));
+    // dbgln("block length symbol: {}", symbol);
+
+    size_t block_length = block_length_code_base[symbol] + TRY(m_input_stream.read_bits(block_length_code_extra[symbol]));
```

Comment:
These buffers can be overflowed:
[crash-f87d381e0dcfe2928983707d18780486c71c24d7.txt](https://github.com/SerenityOS/serenity/files/8421905/crash-f87d381e0dcfe2928983707d18780486c71c24d7.txt)
[crash-4daea4fa729099ae5db13a334e4f0e9a2bc5a4a6.txt](https://github.com/SerenityOS/serenity/files/8421941/crash-4daea4fa729099ae5db13a334e4f0e9a2bc5a4a6.txt)